java/src/jmri/jmrix/can/cbus/swing/bootloader/CbusBootloaderPane.java

```diff
@@ -50,6 +50,10 @@
 public class CbusBootloaderPane extends jmri.jmrix.can.swing.CanPanel
         implements CanListener {
 
+    protected static final int EE_START_2580 = 0xf00000;
+    protected static final int EE_START_26K83 = 0x310000;
+    protected static final int EE_START_27Q84 = 0x380000;
```

Comment:
I think it would be good to add comments to these constants, for example:

```
// EEPROM base address for PIC 18F27-47-57Q84
protected static final int EE_START_27Q84 = 0x380000;
```

Reply:
Rather than coding microcontroller-specific information into JMRI, it would be better if the firmware file contained the EE_START value for the microcontroller, and any other microcontroller-specific information.  Including that microcontroller-specific information in the microcontroller-specific firmware file would enable JMRI to read the info from the firmware file.  There would be no need to re-code JMRI every time a new CBUS device used a microcontroller not yet comprehended by JMRI.  JMRI coding would only be needed when a new update _procedure_ was required.

LocoNet device firmware files already support that sort of mechanism...

Reply:
@danielb987 Good idea. I changed the format slightly as to cover all of the related devices would become an unwieldy string.

@devel-bobm I understand what you are saying, but the firmware files are the intel hex files output by the PIC development tools. We would need to post-process them to include extra information, in which case they would no longer be pure hex files and would (possibly) break other tools, as well as JMRI.

The bootloader client can determine the device type (but not, directly, address boundaries) from the module being updated. An easily updated xml file might be a better way to store these mappings. Longer term I intend to merge the bootloader with the Node Manager and will consider changes then. Also longer term we need a better bootloader protocol and will look at making it device agnostic, e.g., simply pipe the hex file to the module and let it do the parsing.

Reply:
@crosland,

>the firmware files are the intel hex files output by the PIC development tools. We would need to post-process them to include extra information, in which case they would no longer be pure hex files and would (possibly) break other tools, as well as JMRI.

None of these have been hurdles for Dick Brosnan of RR-CirKits, for his firmware-updatable LocoNet products.

The trick is to place the microprocessor-specific metadata into _comments_ in the '.hex' file, with a well-defined structure that JMRI can expect the firmware to follow.  Comments are already supported by the '.hex' file format(s).  Pre-pending a few product-specific comment lines to the generated '.hex' file is _really_ simple, and should not be considered a barrier to such a solution.  Especially for developers who code using the Netbeans-based Microchip MPLAB product, where ant and/or maven can define the build process, including post-processing steps. 

As far as compatibility with old firmware files: if the '.hex' file does not have the metadata, then the _existing_ JMRI code uses the microcontroller identity from the firmware file to look-up the appropriate (coded) values to fill in its variables.  If the metadata is present in the '.hex', then JMRI code extracts the info from the comments in the file to fill in its variables.  If all the required variables are initialized to non-default, valid values, then the firmware update process can proceed.  If any variables have not been configured, or are at illegal values, the firmware update process cannot proceed.

Such a procedure would be particularly well suited to the "better bootloader protocol" you mention.

I personally prefer for hardware-specific things to be handled via the hardware-specific features.  And you have to admit that a firmware '.hex' file is very hardware-specific.  That makes the '.hex' file the _perfect_ place to hold the metadata, instead of in any JMRI file or JMRI code!

Reply:
Yes, I understand, and mostly agree with what you say, but I don't have control over all the non-JMRI tools and don't know that they will not barf over comments in the hex files. The very simple hex file reader I wrote for JMRI will not recognise comments :( I can also see resistance from MERG members to any need to fix up hex files, some of which are still produced with MPLAB 8 which is not ant or maven based.